In [14]:
import pandas as pd


In [15]:
# criando datas sets a partir de dicionários

pessoas= pd.DataFrame({
    'Nome':['Ana', 'Carlos','Diego','Samanta'], 
    'Idade':[25,30,21,19]
    })

#exibindo data frame

pessoas

Nome  Idade
0      Ana     25
1   Carlos     30
2    Diego     21
3  Samanta     19

In [16]:
#Criando Dataframes a partir de Listas

pessoas_02= pd.DataFrame([
    ['Ana', 15],
    ['Eduardo',40],
    ['Diego',28],
    ['Carol', 29]
    ],
    columns= ['Nome','Idade'])

#Exibindo Dataframe

pessoas_02

Nome  Idade
0      Ana     15
1  Eduardo     40
2    Diego     28
3    Carol     29

In [17]:
df.tail() #5 ultimas linhas

age     sex    bmi  children smoker     region     charges
1333   50    male  30.97         3     no  northwest  10600.5483
1334   18  female  31.92         0     no  northeast   2205.9808
1335   18  female  36.85         0     no  southeast   1629.8335
1336   21  female  25.80         0     no  southwest   2007.9450
1337   61  female  29.07         0    yes  northwest  29141.3603

In [18]:
#importando arquivos

df = pd.read_csv("Insurance.csv")

df.head() #exibe 5 primeiras linhas

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [19]:
df.shape #exibe quantidade de linhas e colunas no dataframe, o certo é chamar indice e colunas

(1338, 7)

In [20]:
df.info() #exibe infomações gerais do data set e o tipo de dados

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


In [21]:
#Podemos ter como boa pratica exibir o seguinte print para facilitar quem vai ler nossos trabalhos

print(f' Temos {df.shape[0]} linhas e {df.shape[1]} colunas')

 Temos 1338 linhas e 7 colunas


In [22]:
#podemos acessar as colunas de duas maneiras, df['namecolumn'] or df.namecolumn

#Usando colchetes
print (df['age'])

#usando notação de ponto

print(df.age)

#ambas possuiem a mesma saída, e mostraram as 5 primeiras linhas e as últimas 5 linhas do dataframe

0       19
1       18
2       28
3       33
4       32
        ..
1333    50
1334    18
1335    18
1336    21
1337    61
Name: age, Length: 1338, dtype: int64
0       19
1       18
2       28
3       33
4       32
        ..
1333    50
1334    18
1335    18
1336    21
1337    61
Name: age, Length: 1338, dtype: int64


Podemos também delimitar a quantidade de linhas ou observações que queremos analisar da seguinte forma:

In [23]:
#exibindo a linha 3 a 8
df['age'][3:8]

3    33
4    32
5    31
6    46
7    37
Name: age, dtype: int64

In [24]:
#exibindo a linha de 3 a 8
df.age[3:8]

3    33
4    32
5    31
6    46
7    37
Name: age, dtype: int64

Quando queremos selecionar mais de uma coluna no DATAFRAME, precisamos passar por um formato de lista, assim o pandas irá entender que se trata de colunas diferentes. Tipo, ['idade', 'nome'], 'etc']. No método do Ponto só podemos acessar uma coluna por vez, por isso acabamos utilizando essa forma.

In [25]:
df[['age','sex', 'bmi']]

age     sex     bmi
0      19  female  27.900
1      18    male  33.770
2      28    male  33.000
3      33    male  22.705
4      32    male  28.880
...   ...     ...     ...
1333   50    male  30.970
1334   18  female  31.920
1335   18  female  36.850
1336   21  female  25.800
1337   61  female  29.070

[1338 rows x 3 columns]

#### ACESSAR LINHAS ####

In [26]:
df.loc[0]

age                19
sex            female
bmi              27.9
children            0
smoker            yes
region      southwest
charges     16884.924
Name: 0, dtype: object

In [29]:
#podemos selecionar linhas por meio de listas
df.loc[[0,2]] #seleciona a linha de 0 e 2

age     sex   bmi  children smoker     region    charges
0   19  female  27.9         0    yes  southwest  16884.924
2   28    male  33.0         3     no  southeast   4449.462

In [30]:
#selecionar linha de uma columana específica
df.loc[0,'bmi']

27.9

In [31]:
#Podemos selecionar células de acordo com condicoes

#tipo, quem tem mais de 63 anos

df.loc[df['age']>63]

age     sex     bmi  children smoker     region      charges
62     64    male  24.700         1     no  northwest  30166.61817
94     64  female  31.300         2    yes  southwest  47291.05500
199    64  female  39.330         0     no  northeast  14901.51670
328    64  female  33.800         1    yes  southwest  47928.03000
335    64    male  34.500         0     no  southwest  13822.80300
378    64  female  30.115         3     no  northwest  16455.70785
398    64    male  25.600         2     no  southwest  14988.43200
402    64  female  32.965         0     no  northwest  14692.66935
418    64    male  39.160         1     no  southeast  14418.28040
420    64    male  33.880         0    yes  southeast  46889.26120
534    64    male  40.480         0     no  southeast  13831.11520
603    64  female  39.050         3     no  southeast  16085.12750
635    64    male  38.190         0     no  northeast  14410.93210
664    64  female  22.990         0    yes  southeast  27037.91410
752    64    male  37.905         0     no  northwest  14210.53595
768    64  female  39.700         0     no  southwest  14319.03100
801    64  female  35.970         0     no  southeast  14313.84630
825    64  female  31.825         2     no  northeast  16069.08475
890    64  female  26.885         0    yes  northwest  29330.98315
1051   64    male  26.410         0     no  northeast  14394.55790
1241   64    male  36.960         2    yes  southeast  49577.66240
1265   64    male  23.760         0    yes  southeast  26926.51440

In [32]:
#também aceita multiplas condições tal
#seleciona homens acima de 30 anos

df.loc[((df['age']>30)&(df['sex']=='male'))]

age   sex     bmi  children smoker     region      charges
3      33  male  22.705         0     no  northwest  21984.47061
4      32  male  28.880         0     no  northwest   3866.85520
8      37  male  29.830         2     no  northeast   6406.41070
18     56  male  40.300         0     no  southwest  10602.38500
24     37  male  28.025         2     no  northwest   6203.90175
...   ...   ...     ...       ...    ...        ...          ...
1324   31  male  25.935         1     no  northwest   4239.89265
1325   61  male  33.535         0     no  northeast  13143.33665
1327   51  male  30.030         1     no  southeast   9377.90470
1329   52  male  38.600         2     no  southwest  10325.20600
1333   50  male  30.970         3     no  northwest  10600.54830

[446 rows x 7 columns]

#### manipulação de dados ####

Também podemos adicionar e excluir novas clunas, modificar valores e calcular estatísticas, nesse caso iremos pegar o df para o ano de 2023 e adicionar uma nova coluna com uma condição


In [35]:
import numpy as np

In [36]:
#adicionando colunas
df['Ano_Analise'] =2023

df['Cobrança_Abusiva']= np.where(
    df['charges'] > 20000,
    "Sim",
    "Não"
)

df.head()

age     sex     bmi  children smoker     region      charges  Ano_Analise  \
0   19  female  27.900         0    yes  southwest  16884.92400         2023   
1   18    male  33.770         1     no  southeast   1725.55230         2023   
2   28    male  33.000         3     no  southeast   4449.46200         2023   
3   33    male  22.705         0     no  northwest  21984.47061         2023   
4   32    male  28.880         0     no  northwest   3866.85520         2023   

  Cobrança_Abusiva  
0              Não  
1              Não  
2              Não  
3              Sim  
4              Não

A função Where do numpy funciona como o if e elfse, ela irá verificar se uma condição é verdade, caso for verdadeiro, ela retornará 'Sim', caso contrário, ela retornará 'Não'. Nesse caso em específico, ela está verificando se a coluna charges é > 20000

 <b> Excluir colunas </b>

 Podemos utilizar o <b>df.drop ou inplace </b>. Qual a diferença? No df.drop ele irá apenas ocultar a visualização, mas não irá alterar o objeto no dataframe original. Por isso, é importante utilizar o inplace quando de fato você queira excluir definitivamente alguma informação.

In [37]:
df.drop(columns="Ano_Analise")

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

     Cobrança_Abusiva  
0                 Não  
1                 Não  
2                 Não  
3                 Sim  
4                 Não  
...               ...  
1333              Não  
1334              Não  
1335              Não  
1336              Não  
1337              Sim  

[1338 rows x 8 columns]

tipo <br>
df.drop(columns="Ano_Analise", inplace=True)


Para atribuir novos valores, podemos utilizar a função map, nesse caso iremos substituir a coluna do dataframe que criamos momentos atrás, chamada cobrança abusiva e iremos susbtituir o Sim e o Não, por 1 e 0, da seguinte forma:


In [46]:
#trocando sim por 1 e não por 0
df['Cobrança_Abusiva'] = df['Cobrança_Abusiva'].map({
    "Sim":1,
    "Não":0
})

#criando coluna com valor 2023

df["Ano_Analise"] = 2023

#Atualiza valor de 2023 para 2022

df["Ano_Analise"] = 2022

#exibindo

df.head()

age     sex     bmi  children smoker     region      charges  \
0   19  female  27.900         0    yes  southwest  16884.92400   
1   18    male  33.770         1     no  southeast   1725.55230   
2   28    male  33.000         3     no  southeast   4449.46200   
3   33    male  22.705         0     no  northwest  21984.47061   
4   32    male  28.880         0     no  northwest   3866.85520   

   Cobrança_Abusiva  Ano_Analise  
0               NaN         2022  
1               NaN         2022  
2               NaN         2022  
3               NaN         2022  
4               NaN         2022

In [47]:
print(df['Cobrança_Abusiva'].unique())

[nan]


In [48]:
df['Cobrança_Abusiva'].fillna("Não Informado", inplace=True)
df['Cobrança_Abusiva'] = df['Cobrança_Abusiva'].map({
    "Sim": 1,
    "Não": 0,
    "Não Informado": 0  # ou qualquer valor padrão que faça sentido no seu contexto
})
#criando coluna com valor 2023

df["Ano_Analise"] = 2023

#Atualiza valor de 2023 para 2022

df["Ano_Analise"] = 2022

#exibindo

df.head()

age     sex     bmi  children smoker     region      charges  \
0   19  female  27.900         0    yes  southwest  16884.92400   
1   18    male  33.770         1     no  southeast   1725.55230   
2   28    male  33.000         3     no  southeast   4449.46200   
3   33    male  22.705         0     no  northwest  21984.47061   
4   32    male  28.880         0     no  northwest   3866.85520   

   Cobrança_Abusiva  Ano_Analise  
0                 0         2022  
1                 0         2022  
2                 0         2022  
3                 0         2022  
4                 0         2022

#### ANALISANDO DADOS ####



In [49]:
#descreve os dados do dataframe
df.describe()

age          bmi     children       charges  Cobrança_Abusiva  \
count  1338.000000  1338.000000  1338.000000   1338.000000            1338.0   
mean     39.207025    30.663397     1.094918  13270.422265               0.0   
std      14.049960     6.098187     1.205493  12110.011237               0.0   
min      18.000000    15.960000     0.000000   1121.873900               0.0   
25%      27.000000    26.296250     0.000000   4740.287150               0.0   
50%      39.000000    30.400000     1.000000   9382.033000               0.0   
75%      51.000000    34.693750     2.000000  16639.912515               0.0   
max      64.000000    53.130000     5.000000  63770.428010               0.0   

       Ano_Analise  
count       1338.0  
mean        2022.0  
std            0.0  
min         2022.0  
25%         2022.0  
50%         2022.0  
75%         2022.0  
max         2022.0

<p><b>count:</b> Mostra o número de observações (linhas) não-nulas em cada coluna do DataFrame.</p>
<p><b>mean:</b> Calcula a média aritmética dos valores em cada coluna. É a soma de todos os valores dividida pelo número de valores.</p>
<p><b>std:</b> Representa o desvio padrão, que mede a quantidade de variação ou dispersão de um conjunto de valores. Um desvio padrão baixo indica que os valores tendem a estar próximos da média; um desvio padrão alto indica que os valores estão espalhados por uma gama mais ampla de valores.</p>
<p><b>min:</b> Mostra o valor mínimo em cada coluna do DataFrame.</p>
<p><b>25%:</b> Representa o primeiro quartil (Q1) dos dados, ou seja, 25% dos dados são menores que esse valor. É calculado ordenando os dados e tomando o valor no índice que é 25% do caminho através da lista.</p>
<p><b>50%:</b> Representa a mediana ou o segundo quartil (Q2) dos dados. Metade dos dados são menores que esse valor. É o valor no meio de um conjunto de dados ordenado.</p>
<p><b>75%:</b> Representa o terceiro quartil (Q3) dos dados, ou seja, 75% dos dados são menores que esse valor. É o valor abaixo do qual se encontram 75% dos dados.</p>
<p><b>max:</b> Mostra o valor máximo em cada coluna do DataFrame.</p>

<p>Ou seja, ao olharmos para as idades (age), temos:</p>
<ul>
    <li>1338 observações não-nulas. Como o DataFrame contém 1338 linhas, não há idade missing aqui;</li>
    <li>Média de idade de aproximadamente 39.21 anos;</li>
    <li>Um desvio padrão de 14, parece haver uma dispersão considerável na idade. I.e., temos bastantes pessoas com menos de 30 e mais de 50;</li>
    <li>Um Q1 de 27 anos. Ou seja, 25% das observações possuem menos de 27 anos;</li>
    <li>50% das observações possuem menos de 39 anos. Como nossa média é 39.21, parece que temos dados sem muita assimetria à direita ou à esquerda;</li>
    <li>75% das observações possuem menos de 51 anos;</li>
    <li>A pessoa mais velha do DataFrame possui 64 anos.</li>
</ul>

<p>O <code>describe()</code> é ótimo para demonstrar um resumo geral, mas apenas para colunas numéricas. Se quisermos olhar as categóricas, vamos utilizar <code>value_counts()</code>:</p>

<p> Para fechar, e se quisermos cruzar as variáveis? Por exemplo, podemos querer saber se a média de idade é diferente para homens ou mulheres: </p>


In [51]:
df.groupby("sex")["age"].mean()

sex
female    39.503021
male      38.917160
Name: age, dtype: float64

<p> Estamos agrupando a idade por sexo nesse caso e tiramos a média pela função mean() </p>

In [52]:
#desvio padrão por sexo e fumante

df.groupby(['smoker','sex'])['charges'].std()

smoker  sex   
no      female     6060.775970
        male       5908.108989
yes     female    11907.536381
        male      11202.670862
Name: charges, dtype: float64